File: src/desktop-edge-agent/src/FccDesktopAgent.Core/Adapter/Doms/Protocol/DomsTransactionParser.cs
using FccDesktopAgent.Core.Adapter.Doms.Jpl;
using FccDesktopAgent.Core.Adapter.Doms.Model;

namespace FccDesktopAgent.Core.Adapter.Doms.Protocol;

/// <summary>
/// Handles the DOMS supervised transaction buffer: lock -> read -> clear.
///
/// DOMS maintains a "supervised buffer" of completed transactions per fuelling point.
/// To safely retrieve transactions:
///   1. Lock the buffer (FpSupTrans_lock_req) -- prevents new transactions from entering
///   2. Read transactions (FpSupTrans_read_req) -- returns buffered transactions
///   3. Clear the buffer (FpSupTrans_clear_req) -- removes read transactions
///
/// If clear is not sent, transactions remain in the buffer and will be returned again.
/// </summary>
public static class DomsTransactionParser
{
    // -- JPL message names -------------------------------------------------------

    public const string LockRequest = "FpSupTrans_lock_req";
    public const string LockResponse = "FpSupTrans_lock_resp";
    public const string ReadRequest = "FpSupTrans_read_req";
    public const string ReadResponse = "FpSupTrans_read_resp";
    public const string ClearRequest = "FpSupTrans_clear_req";
    public const string ClearResponse = "FpSupTrans_clear_resp";

    /// <summary>Result code indicating success.</summary>
    public const string ResultOk = "0";

    /// <summary>Result code indicating no transactions available.</summary>
    public const string ResultEmpty = "1";

    // -- Lock --------------------------------------------------------------------

    /// <summary>
    /// Build a lock request for the supervised transaction buffer.
    /// </summary>
    /// <param name="fpId">Fuelling point ID (0 = all pumps).</param>
    public static JplMessage BuildLockRequest(int fpId = 0)
    {
        return new JplMessage(
            Name: LockRequest,
            Data: new Dictionary<string, string>
            {
                ["FpId"] = fpId.ToString(),
            });
    }

    /// <summary>
    /// Validate a lock response.
    /// </summary>
    /// <returns>true if lock was acquired successfully.</returns>
    public static bool ValidateLockResponse(JplMessage response)
    {
        if (response.Name != LockResponse)
            return false;

        if (response.Data is null || !response.Data.TryGetValue("ResultCode", out var resultCode))
            return false;

        return resultCode == ResultOk;
    }

    // -- Read --------------------------------------------------------------------

    /// <summary>
    /// Build a read request for the supervised transaction buffer.
    /// </summary>
    /// <param name="fpId">Fuelling point ID (0 = all pumps).</param>
    /// <param name="bufferIndex">Starting buffer index to read from.</param>
    public static JplMessage BuildReadRequest(int fpId = 0, int bufferIndex = 0)
    {
        return new JplMessage(
            Name: ReadRequest,
            Data: new Dictionary<string, string>
            {
                ["FpId"] = fpId.ToString(),
                ["BufferIndex"] = bufferIndex.ToString(),
            });
    }

    /// <summary>
    /// Parse a read response into transaction DTOs.
    /// </summary>
    /// <param name="response">JPL response message.</param>
    /// <returns>List of parsed transactions, or empty if buffer was empty.</returns>
    public static IReadOnlyList<DomsJplTransactionDto> ParseReadResponse(JplMessage response)
    {
        if (response.Name != ReadResponse)
            return [];

        if (response.Data is null)
            return [];

        var data = response.Data;

        if (!data.TryGetValue("ResultCode", out var resultCode))
            return [];

        if (resultCode == ResultEmpty)
            return [];

        if (resultCode != ResultOk)
            return [];

        if (!data.TryGetValue("TransCount", out var countStr) || !int.TryParse(countStr, out var count))
            return [];

        if (count == 0)
            return [];

        // For a single transaction, parse directly from the data map.
        // For multiple transactions, data contains indexed fields (Trans_0_*, Trans_1_*, etc.).
        if (count == 1)
        {
            return [DomsSupParamParser.Parse(data, bufferIndex: 0)];
        }

        var results = new List<DomsJplTransactionDto>(count);
        for (var index = 0; index < count; index++)
        {
            try
            {
                var indexedData = ExtractIndexedTransaction(data, index);
                results.Add(DomsSupParamParser.Parse(indexedData, bufferIndex: index));
            }
            catch
            {
                // Skip malformed individual transactions.
            }
        }

        return results;
    }

    // -- Clear -------------------------------------------------------------------

    /// <summary>
    /// Build a clear request to remove read transactions from the buffer.
    /// </summary>
    /// <param name="fpId">Fuelling point ID.</param>
    /// <param name="count">Number of transactions to clear.</param>
    public static JplMessage BuildClearRequest(int fpId = 0, int count = 0)
    {
        return new JplMessage(
            Name: ClearRequest,
            Data: new Dictionary<string, string>
            {
                ["FpId"] = fpId.ToString(),
                ["TransCount"] = count.ToString(),
            });
    }

    /// <summary>
    /// Validate a clear response.
    /// </summary>
    /// <returns>true if clear was successful.</returns>
    public static bool ValidateClearResponse(JplMessage response)
    {
        if (response.Name != ClearResponse)
            return false;

        if (response.Data is null || !response.Data.TryGetValue("ResultCode", out var resultCode))
            return false;

        return resultCode == ResultOk;
    }

    // -- Helpers -----------------------------------------------------------------

    /// <summary>
    /// Extract indexed transaction fields from a multi-transaction response.
    /// Fields are prefixed with "Trans_{index}_" (e.g., "Trans_0_TransId", "Trans_0_FpId").
    /// </summary>
    private static Dictionary<string, string> ExtractIndexedTransaction(
        IReadOnlyDictionary<string, string> data, int index)
    {
        var prefix = $"Trans_{index}_";
        var result = new Dictionary<string, string>();

        foreach (var kvp in data)
        {
            if (kvp.Key.StartsWith(prefix, StringComparison.Ordinal))
            {
                result[kvp.Key[prefix.Length..]] = kvp.Value;
            }
        }

        return result;
    }
}

File: src/edge-agent/app/src/main/kotlin/com/fccmiddleware/edge/adapter/doms/protocol/DomsTransactionParser.kt
package com.fccmiddleware.edge.adapter.doms.protocol

import com.fccmiddleware.edge.adapter.doms.jpl.JplMessage
import com.fccmiddleware.edge.adapter.doms.model.DomsTransactionDto

/**
 * Handles the DOMS supervised transaction buffer: lock → read → clear.
 *
 * DOMS maintains a "supervised buffer" of completed transactions per fuelling point.
 * To safely retrieve transactions:
 *   1. Lock the buffer (FpSupTrans_lock_req) — prevents new transactions from entering
 *   2. Read transactions (FpSupTrans_read_req) — returns buffered transactions
 *   3. Clear the buffer (FpSupTrans_clear_req) — removes read transactions
 *
 * If clear is not sent, transactions remain in the buffer and will be returned again.
 */
object DomsTransactionParser {

    // ── JPL message names ────────────────────────────────────────────────────

    const val LOCK_REQUEST = "FpSupTrans_lock_req"
    const val LOCK_RESPONSE = "FpSupTrans_lock_resp"
    const val READ_REQUEST = "FpSupTrans_read_req"
    const val READ_RESPONSE = "FpSupTrans_read_resp"
    const val CLEAR_REQUEST = "FpSupTrans_clear_req"
    const val CLEAR_RESPONSE = "FpSupTrans_clear_resp"

    /** Result code indicating success. */
    const val RESULT_OK = "0"

    /** Result code indicating no transactions available. */
    const val RESULT_EMPTY = "1"

    // ── Lock ─────────────────────────────────────────────────────────────────

    /**
     * Build a lock request for the supervised transaction buffer.
     *
     * @param fpId Fuelling point ID (0 = all pumps).
     */
    fun buildLockRequest(fpId: Int = 0): JplMessage {
        return JplMessage(
            name = LOCK_REQUEST,
            data = mapOf("FpId" to fpId.toString()),
        )
    }

    /**
     * Validate a lock response.
     *
     * @return true if lock was acquired successfully.
     */
    fun validateLockResponse(response: JplMessage): Boolean {
        if (response.name != LOCK_RESPONSE) return false
        val resultCode = response.data["ResultCode"] ?: return false
        return resultCode == RESULT_OK
    }

    // ── Read ─────────────────────────────────────────────────────────────────

    /**
     * Build a read request for the supervised transaction buffer.
     *
     * @param fpId Fuelling point ID (0 = all pumps).
     * @param bufferIndex Starting buffer index to read from.
     */
    fun buildReadRequest(fpId: Int = 0, bufferIndex: Int = 0): JplMessage {
        return JplMessage(
            name = READ_REQUEST,
            data = mapOf(
                "FpId" to fpId.toString(),
                "BufferIndex" to bufferIndex.toString(),
            ),
        )
    }

    /**
     * Parse a read response into transaction DTOs.
     *
     * @param response JPL response message.
     * @return List of parsed transactions, or empty if buffer was empty.
     */
    fun parseReadResponse(response: JplMessage): List<DomsTransactionDto> {
        if (response.name != READ_RESPONSE) return emptyList()

        val resultCode = response.data["ResultCode"]
        if (resultCode == RESULT_EMPTY) return emptyList()
        if (resultCode != RESULT_OK) return emptyList()

        val count = response.data["TransCount"]?.toIntOrNull() ?: return emptyList()
        if (count == 0) return emptyList()

        // For a single transaction, parse directly from the data map
        // For multiple transactions, data contains indexed fields (Trans_0_*, Trans_1_*, etc.)
        return if (count == 1) {
            listOf(DomsSupParamParser.parse(response.data, bufferIndex = 0))
        } else {
            (0 until count).mapNotNull { index ->
                try {
                    val indexedData = extractIndexedTransaction(response.data, index)
                    DomsSupParamParser.parse(indexedData, bufferIndex = index)
                } catch (_: Exception) {
                    null
                }
            }
        }
    }

    // ── Clear ────────────────────────────────────────────────────────────────

    /**
     * Build a clear request to remove read transactions from the buffer.
     *
     * @param fpId Fuelling point ID.
     * @param count Number of transactions to clear.
     */
    fun buildClearRequest(fpId: Int = 0, count: Int): JplMessage {
        return JplMessage(
            name = CLEAR_REQUEST,
            data = mapOf(
                "FpId" to fpId.toString(),
                "TransCount" to count.toString(),
            ),
        )
    }

    /**
     * Validate a clear response.
     *
     * @return true if clear was successful.
     */
    fun validateClearResponse(response: JplMessage): Boolean {
        if (response.name != CLEAR_RESPONSE) return false
        val resultCode = response.data["ResultCode"] ?: return false
        return resultCode == RESULT_OK
    }

    // ── Helpers ──────────────────────────────────────────────────────────────

    /**
     * Extract indexed transaction fields from a multi-transaction response.
     * Fields are prefixed with "Trans_{index}_" (e.g., "Trans_0_TransId", "Trans_0_FpId").
     */
    private fun extractIndexedTransaction(data: Map<String, String>, index: Int): Map<String, String> {
        val prefix = "Trans_${index}_"
        return data.entries
            .filter { it.key.startsWith(prefix) }
            .associate { it.key.removePrefix(prefix) to it.value }
    }
}
